File: Programmers/BFS/경주로 건설.cpp
// 프로그래머스 경주로 건설 BFS
#include <string>
#include <vector>
#include <queue>
#include <tuple>
#include <climits>

using namespace std;

int n;
int dx[] = {0, -1, 0, 1};
int dy[] = {-1, 0, 1, 0};
vector<vector<vector<int>>> visited;
void BFS(int x, int y, vector<vector<int>> board) {
    queue<tuple<int, int, int, int>> q;
    q.push({x, y, -1, 0}); // 초기에 방향 설정 X

    while(!q.empty()) {
        auto [x, y, last, cost] = q.front();
        q.pop();

        for (int i = 0; i < 4; i++) {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (nx < 0 || n <= nx || ny < 0 || n <= ny) continue;

            int newCost = cost + ((last == -1 || last == i) ? 100 : 600);
            if (visited[nx][ny][i] > newCost && board[nx][ny] == 0) {
                visited[nx][ny][i] = newCost;
                q.push({nx, ny, i, newCost});
            }
        }
    }
}
int solution(vector<vector<int>> board) {
    n = board.size();
    visited.assign(n, vector<vector<int>>(n, vector<int>(4, INT_MAX)));

    BFS(0, 0, board);

    int r = INT_MAX;
    for (int i = 0; i < 4; i++) {
        r = min(r, visited[n - 1][n - 1][i]);
    }
    return r;
}

File: Programmers/BFS/경주로 건설.py
from collections import deque

dx = [0, -1, 0, 1]
dy = [-1, 0, 1, 0]

def bfs(n, x, y, board, visited):
    queue = deque()
    queue.append((x, y, -1, 0))

    while queue:
        x, y, last, cost = queue.popleft()

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if nx < 0 or n <= nx or ny < 0 or n <= ny: continue
            # 회전(500) + 다음칸(100)
            newCost = cost + 100 if last == -1 or last == i else cost + 600

            # 이미 방문했더라도 더 작은 Cost의 경로가 있다면 방문
            if visited[nx][ny][i] > newCost and board[nx][ny] == 0:
                visited[nx][ny][i] = newCost
                queue.append((nx, ny, i, newCost))

def solution(board):
    answer = 0
    n = len(board)
    visited = [[[float('inf') for _ in range(4)] for _ in range(n)] for _ in range(n)]

    bfs(n, 0, 0, board, visited)
    answer = min(visited[n-1][n-1])
    return answer